tests/benchmarks/timing_info/src/arch/x86/tsc.c

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <zephyr.h>
+#include <tc_util.h>
+#include <ksched.h>
+#include <sys_clock.h>
+
+#define CALIB_LOOPS	8
+
+static u64_t tsc_freq;
+
+void calibrate_timer(void)
+{
+	u64_t tsc_start, tsc_end;
+	int i;
+
+	tsc_freq = 0;
+
+	for (i = 0; i < CALIB_LOOPS; i++) {
+		tsc_start = z_tsc_read();
+
+		k_busy_wait(1 * USEC_PER_MSEC);
+
+		tsc_end = z_tsc_read();
+
+		tsc_freq += tsc_end - tsc_start;
+	}
```

Comment:
Not sure I see the point of that loop.  You're incuring measurement
error each time (i.e. almost certainly spending a little longer than a
millisecond between the two TSC reads), and averaging 8 runs doesn't
do anything to address that, becuase the error isn't random.

Both HPET and TSC counters are deterministic, free-running and not
related to interrupt or CPU behavior.  Just let them both run for
however long you want to calibrate for and check them at the
boundaries.

Something like this should be simpler and more accurate:


```c
u32_t cyc_start = k_cycle_get_32();
u64_t tsc_start = z_tsc_read();

k_busy_wait(8000); /* Or anything else to waste time */

u32_t cyc_end = k_cycle_get_32();
u64_t tsc_end = z_tsc_read();

u64_t cyc_freq = sys_clock_hw_cycles_per_sec();
u64_t dcyc = cyc_end - cyc_start;
u64_t dtsc = tsc_end - tsc_start;

u64_t tsc_freq = (cyc_freq * dtsc) / dcyc;
	
```

Reply:
Sorry, note some subtlety: the 32 bit cycle result is assigned to a u64_t but MUST be calculated in 32 bit precision, otherwise the wraparound will be wrong (i.e. it will wrap to ~0xffffffffffff0000 or whatever).

Reply:
Thanks. Fixed.